dagql/server.go

```diff
@@ -0,0 +1,890 @@
+	if sel, ok := val.(Object); ok {
+		// We always support returning something that's already Selectable, e.g. an
+		// object loaded from its ID.
+		return sel, nil
+	}
+	class, ok := s.objects[val.Type().Name()]
+	if !ok {
+		return nil, fmt.Errorf("toSelectable: unknown type %q", val.Type().Name())
+	}
+	return class.New(chainedID, val)
+}
+
+func (s *Server) parseASTSelections(ctx context.Context, gqlOp *graphql.OperationContext, self *ast.Type, astSels ast.SelectionSet) ([]Selection, error) {
+	vars := gqlOp.Variables
+
+	class := s.objects[self.Name()]
+	if class == nil {
+		return nil, fmt.Errorf("parseASTSelections: not an Object type: %q", self.Name())
+	}
+
+	sels := []Selection{}
+	for _, sel := range astSels {
+		switch x := sel.(type) {
+		case *ast.Field:
+			sel, resType, err := class.ParseField(ctx, x, vars)
+			if err != nil {
+				return nil, fmt.Errorf("parse field %q: %w", x.Name, err)
+			}
+			var subsels []Selection
+			if len(x.SelectionSet) > 0 {
+				subsels, err = s.parseASTSelections(ctx, gqlOp, resType, x.SelectionSet)
+				if err != nil {
+					return nil, err
+				}
+			}
+			sels = append(sels, Selection{
+				Alias:         x.Alias,
+				Selector:      sel,
+				Subselections: subsels,
+			})
+		case *ast.FragmentSpread:
+			fragment := gqlOp.Doc.Fragments.ForName(x.Name)
+			if fragment == nil {
+				return nil, fmt.Errorf("unknown fragment: %s", x.Name)
+			}
+			if len(fragment.SelectionSet) > 0 {
+				subsels, err := s.parseASTSelections(ctx, gqlOp, self, fragment.SelectionSet)
+				if err != nil {
+					return nil, err
+				}
+				sels = append(sels, subsels...)
+			}
+		default:
+			return nil, fmt.Errorf("unknown field type: %T", x)
+		}
+	}
+
+	return sels, nil
+}
+
+// Selection represents a selection of a field on an object.
+type Selection struct {
+	Alias         string
+	Selector      Selector
+	Subselections []Selection
+}
+
+// Name returns the name of the selection, which is either the alias or the
+// field name.
+func (sel Selection) Name() string {
+	if sel.Alias != "" {
+		return sel.Alias
+	}
+	return sel.Selector.Field
+}
+
+// Selector specifies how to retrieve a value from an Instance.
+type Selector struct {
+	Field string
+	Args  []NamedInput
+	Nth   int
+}
+
+func (sel Selector) String() string {
+	str := sel.Field
+	if len(sel.Args) > 0 {
+		str += "("
+		for i, arg := range sel.Args {
+			if i > 0 {
+				str += ", "
+			}
+			str += arg.String()
+		}
+		str += ")"
+	}
+	if sel.Nth != 0 {
+		str += fmt.Sprintf("[%d]", sel.Nth)
+	}
+	return str
+}
+
+func (sel Selector) AppendTo(id *idproto.ID, spec FieldSpec) *idproto.ID {
+	astType := spec.Type.Type()
+	tainted := !spec.Pure
+	idArgs := make([]*idproto.Argument, 0, len(sel.Args))
+	for _, arg := range sel.Args {
+		if arg.Value == nil {
+			// we don't include null arguments, since they would needlessly bust caches
+			continue
+		}
+		if arg, found := spec.Args.Lookup(arg.Name); found && arg.Sensitive {
```

Comment:
I'm guessing this is for the `SetSecret` case? And there's no problems with the value here not being part of the ID? It can still make its way to the actual implementation that stores the secret in the store even when not in the ID?